sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -0,0 +1,510 @@
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr& operator++() noexcept;
+----
+|
+Prefix increment operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr operator++() noexcept;
+----
+|
+Postfix increment operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr& operator--() noexcept;
+----
+|
+Prefix decrement operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr operator--() noexcept;
+----
+|
+Postfix decrement operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename propertyT>
+static constexpr bool has_property();
+----
+| Returns true if the `PropertyListT` contains the property specified by `propertyT`. Returns false if it does not.
+Available only if `sycl::is_property_of_v<propertyT, sycl::ext::oneapi::annotated_ptr>` is true.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename propertyT>
+static constexpr auto get_property();
+----
+| Returns an object of the class used to represent the value of property `propertyT`.
+Must produce a compiler diagnostic if `PropertyListT` does not contain a `propertyT` property.
+Available only if `sycl::is_property_of_v<propertyT, sycl::ext::oneapi::annotated_ptr>` is true.
+|===
+
+=== Add new reference wrapper class `annotated_ref` to enable `annotated_ptr`
+
+The purpose of the `annotated_ref` class template is to provide reference wrapper semantics. It enables the implementation to preserve the properties on loads from and stores to the pointers.
+
+```c++
+namespace sycl::ext::oneapi {
+template <typename T, typename PropertyListT = property_list_t<>>
+class annotated_ref {
+  public:
+    annotated_ref(T *);
+    operator T() noexcept;
+    operator const T() const noexcept;
+    void operator=(const T &);
+  private:
+    T *Ptr;
+  };
+} // namespace sycl::ext::oneapi
+```
+
+
+Member Functions are described in the table below
+[frame="topbot",options="header"]
+|===
+|Functions |Description
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ref(T * InputPtr);
+----
+|
+Constructs an `annotated_ref` object. Does not allocate new storage. The underlying pointer is initialized to `InputPtr`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+operator T() noexcept;
+operator const T() const noexcept;
+----
+|
+Implicit conversion to underlying type.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+void operator=(const T &);
+----
+|
+Enables assignment to the underlying pointer.
+
+|===
+
+=== Properties for `annotated_ptr` variables
+
+See the extension `sycl_ext_oneapi_annotated_ptr_properties` for the full list of supported properties.
+
+== Issues
+
+1) [RESOLVED] Should we allow implicit conversion to base class by default?
+*Ans: No.*
+
+2) [RESOLVED] How do we support `operator->`?
+*Ans: Not with the initial release.*
+
+3) [RESOLVED] Can `sycl::atomic_ref` be used with `annotated_ref`?
+*Ans: Yes. This discussion is an implementation detail discussion and does not impact the spec.*
```

Comment:
Can you remind me why this is an implementation detail and not part of the extension API?  What is the syntax you would use to do an atomic operation on the value referenced by `annotated_ptr`?

Reply:
Hi Greg, I think my earlier answer was incorrect. `atomic_ref` will not work with `annotated_ref` as is since `atomic_ref` restricts the types it can take. If we want, we can create a sycl extension for `atomic_ref` to support `annotated_ref`. The implementation complexity will depend on how we chose to implement `annotated_ref`: via builtins or via pointer annotations.